libs/community/langchain_community/vectorstores/pinecone.py

```diff
@@ -142,26 +144,34 @@ def add_texts(
         for metadata, text in zip(metadatas, texts):
             metadata[self._text_key] = text
 
-        # For loops to avoid memory issues and optimize when using HTTP based embeddings
-        # The first loop runs the embeddings, it benefits when using OpenAI embeddings
-        # The second loops runs the pinecone upsert asynchronously.
-        for i in range(0, len(texts), embedding_chunk_size):
-            chunk_texts = texts[i : i + embedding_chunk_size]
-            chunk_ids = ids[i : i + embedding_chunk_size]
-            chunk_metadatas = metadatas[i : i + embedding_chunk_size]
-            embeddings = self._embed_documents(chunk_texts)
-            async_res = [
-                self._index.upsert(
-                    vectors=batch,
-                    namespace=namespace,
-                    async_req=True,
-                    **kwargs,
-                )
-                for batch in batch_iterate(
-                    batch_size, zip(chunk_ids, embeddings, chunk_metadatas)
-                )
-            ]
-            [res.get() for res in async_res]
+        if async_req:
+            # For loops to avoid memory issues when using HTTP-based embeddings
+            # First loop runs embeddings, benefits when using OpenAI embeddings
+            # The second loops runs the pinecone upsert asynchronously.
+            for i in range(0, len(texts), embedding_chunk_size):
+                chunk_texts = texts[i : i + embedding_chunk_size]
+                chunk_ids = ids[i : i + embedding_chunk_size]
+                chunk_metadatas = metadatas[i : i + embedding_chunk_size]
+                embeddings = self._embed_documents(chunk_texts)
+                async_res = [
+                    self._index.upsert(
+                        vectors=batch,
+                        namespace=namespace,
+                        async_req=async_req,
+                        **kwargs,
+                    )
+                    for batch in batch_iterate(
+                        batch_size, zip(chunk_ids, embeddings, chunk_metadatas)
+                    )
+                ]
+                [res.get() for res in async_res]
+        else:
+            self._index.upsert(
+                vectors=list(zip(ids, self._embed_documents(texts), metadatas)),
```

Comment:
Thoughts on just passing `async_req=async_req` and retain the same chunking logic as for async version?

Reply:
Great minds think alike - that was my initial idea as well! 

It doesn't work because the line `[res.get() for res in async_res]` throws an error, it's not valid for non-async. 

Here's a traceback:

```python
self = <tests.integration_tests.vectorstores.test_pinecone.TestPinecone object at 0x11a2a0bd0>
embedding_openai = OpenAIEmbeddings(client=<openai.resources.embeddings.Embeddings object at 0x11b067850>, async_client=<openai.resources...kip_empty=False, default_headers=None, default_query=None, retry_min_seconds=4, retry_max_seconds=20, http_client=None)

    @pytest.mark.usefixtures("mock_pool_not_supported")
    def test_that_async_freq_false_enabled_singlethreading(
        self, embedding_openai: OpenAIEmbeddings
    ) -> None:
>       Pinecone.from_texts(
            texts=["foo", "bar", "baz"],
            embedding=embedding_openai,
            index_name=index_name,
            async_req=False,
        )

tests/integration_tests/vectorstores/test_pinecone.py:335:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
langchain_community/vectorstores/pinecone.py:451: in from_texts
    pinecone.add_texts(
langchain_community/vectorstores/pinecone.py:166: in add_texts
    [res.get() for res in async_res]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

.0 = <list_iterator object at 0x11c0a5bd0>

>   [res.get() for res in async_res]
E   TypeError: ModelNormal.get() missing 1 required positional argument: 'name'

langchain_community/vectorstores/pinecone.py:166: TypeError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /Users/nelsonauner/tech/langchain/libs/community/langchain_community/vectorstores/pinecone.py(166)<listcomp>()
-> [res.get() for res in async_res]
(Pdb) res
{'upserted_count': 3}
```